### DISCUSSION TITLE
TOC is missing page numbers

### DISCUSSION THREAD

**Discussion by PiotrPanasiuk** [kjambunathan/org-mode-ox-odt#174] [Q&A] [answered]
When i export by ctrl x E and o. o my TOC is generated without page numbers and all entry are hyperlinks ( blue underlined). When I generate i.e. ctrl x E and r p TOC is generated with page numbers.
Do someone knows if there is possibility to create table of contents with page numbers and ox-odt?

**Comment by QiangF** [comment]
Make sure you are not using org-odt from the default org path (M-x locate-library). I have page number with the default setting. Try export this

    #+options:  toc:nil author:nil tex:t latex:t :H 10
    #+title: hello
    #+toc: headlines 2
    #+pagebreak:
    * h1
    * h2
      
Or you can modified the exported file to whatever you lile, and use it as the style file

    #+odt_styles_file: path-to-your-modified-odt

**Reply by PiotrPanasiuk** [reply]
Thank You for help, I have tried both emacs and spacemacs. Unfortunately my pure TOC ODT output look like. Do You know what I am missing? 
 
![obraz](https://user-images.githubusercontent.com/11033936/159693696-4f06f4bd-59e9-410e-a4c4-0e9f5e186611.png)

**Comment by kjambunathan** [comment]
For resolving load path issues see

https://github.com/kjambunathan/org-mode-ox-odt/discussions/133#discussioncomment-2214682


https://github.com/kjambunathan/org-mode-ox-odt/issues/48#issuecomment-441103145

**Comment by kjambunathan** [comment]
This comment will help you figure out which ox-odt you are running.



https://github.com/kjambunathan/org-mode-ox-odt/issues/48#issuecomment-441103145

Remember my exporter is a fork.  I distribute it via my own ELPA.

**Comment by kjambunathan** [comment]
(Replying from phone)

I added a fix few hours ago and you need to upgrade ...

Did you read the yellow box in my screenshot?
https://github.com/kjambunathan/org-mode-ox-odt/issues/175#issuecomment-1076134280

How did you install my change? Search discussions for straght recipe


Are you sure you are running my ox-odt


M-x locate-library ox-odt


M-x list-load-path-shadows

M-x find-library ox-odt.  See which file and directory it gets you to

What paths you see in "Messages" buffer after export

Try M-x load-file if everything else fails

**Reply by PiotrPanasiuk** [reply]
Thank You once again for help. On my previous mail I have engaged my mouth before brain :(, haven't noticed other posts. So when I have implemented Your workaround, export to ODT works just beautiful, however and unfortunately the variable #+odt_preferred_output_format: pdf and export to PDF stopped working. Thank You once again.

**Comment by kjambunathan** [comment]
If NO pdf file is getting produced at all, may be there is no 'soffice'
binary in your 'executable-path'  or the 'soffice' binary is called
somethin else.  Which OS are you running?

If PDF file is produced without page numbers, then you need to run 'update
all" in Libreoffice manually, and then export it through LibreOffice GUI
all by hand.

If you need to do 'update all' automatically install the OXT file in
LibreOffice, and set up 'org-odt-transform-processes'.


See an example.here

https://kjambunathan.gitlab.io/org-mode-ox-odt/Getting-Started-with-ODT-export.html#Configuration

(You do NOT need "Reload" entry but you definitely need "Update All"
entry.  "Update All" by hand or through LibreOffice macro is what
calculates and inserts page numbers)

**Comment by kjambunathan** [comment]
Check the spelling of keyword (Try C-c C-e # o) or remove the double quotes
around  pdf.

 Or you can customize 'org-odt-preferred-output-format' through M-x
customize-variable.

You have a misconfiguration somewhere

**Comment by kjambunathan** [comment]
Use the full path to soffice  binary (or whatever it is called in your OS)


See
https://github.com/takaxp/emacs.d/blob/fe21113b02b5acf8d6d08e4f3d6504ead5dd11c0/init-org.el#L1578

**Reply by PiotrPanasiuk** [reply]
Indeed it was my fault. I have put in wrong position configuration dedicated for ox-odt in .spacemacs init file, when I have corrected it, as bellow, all started working like a charm, and that include pdf generation with correct TOC. Thank You for help.

(custom-set-variables
'(org-odt-convert-process "LibreOffice")
 '(org-odt-transform-processes
   '(("Optimize Column Width of all Tables"
      "soffice" "--norestore" "--invisible" "--headless"
      "macro:///OrgMode.Utilities.OptimizeColumnWidth(%I)")
     ("Update All"
      "soffice" "--norestore" "--invisible" "--headless"
      "macro:///OrgMode.Utilities.UpdateAll(%I)")
     ("Reload"
      "soffice" "--norestore" "--invisible" "--headless"
      "macro:///OrgMode.Utilities.Reload(%I)")))

(setcdr (assq 'system org-file-apps-defaults-gnu) "xdg-open %s")

(advice-add 'org-open-file :around
            (lambda (orig-fun &rest args)
              ;; Work around a weird problem with xdg-open.
              (let ((process-connection-type nil))
                (apply orig-fun args))))

**Reply by kjambunathan** [reply]
> when I have corrected it, as bellow, all started working like a charm, and that include pdf generation with correct TOC.

The documentation is a bit out of date.

You can safely remove `"Optimize Column Width of all Tables"` and  `"Reload"` entries in `org-odt-transform-processes`.  That is, you can do something  like what you see below.

The `Reload` is better done as part of a user-specific `*.desktop` files as documented in [Howto make `LibreOffice` reload files on Export](https://github.com/kjambunathan/org-mode-ox-odt/issues/111#issuecomment-989495214).  Using custom `desktop` file with `Reload` is good to have, but not necessary. 

```elisp
(custom-set-variables
 '(org-odt-convert-process "LibreOffice")
 '(org-odt-preferred-output-format "docx")
 '(org-odt-transform-processes
   '(("Update All"
      "soffice" "--norestore" "--invisible" "--headless"
      "macro:///OrgMode.Utilities.UpdateAll(%I)")))
 '(org-latex-to-mathml-convert-command
   "java -jar %j -unicode -force -df %o %I")
 '(org-latex-to-mathml-jar-file
   "/home/kjambunathan/Downloads/mathtoweb.jar"))

```

**Comment by kjambunathan** [comment, marked as answer]
> When i export by ctrl x E and o. o my TOC is generated without page numbers and all entry are hyperlinks ( blue underlined).
> Do someone knows if there is possibility to create table of contents with page numbers and ox-odt?

**_(Marking this discussion as `answered`)_**

This is fixed as part of [TOC is missing page numbers](https://github.com/kjambunathan/org-mode-ox-odt/issues/175).

**Comment by kjambunathan** [comment]
(Replying from phone)

Did you read the yellow box in my screenshot?

How did you install my change? Search discussions for straght recipe


Are you sure you are running my ox-odt

M-x list-load-path-shadows

M-x find-library ox-odt.  See which file gets you to

What paths you see in "Messages" buffer after export

Try M-x load-file if everything else fails
